hal-x86_64/src/mm/mod.rs

```diff
@@ -0,0 +1,831 @@
+    ptr::NonNull,
+    sync::atomic::{AtomicUsize, Ordering},
+};
+use hal_core::{
+    mem::page::{
+        self, Map, Page, Size, TranslateAddr, TranslateError, TranslatePage, TranslateResult,
+    },
+    Address,
+};
+
+const ENTRIES: usize = 512;
+
+type VirtPage<S> = Page<VAddr, S>;
+type PhysPage<S> = Page<PAddr, S>;
+
+pub struct PageCtrl {
+    pml4: NonNull<PageTable<level::Pml4>>,
+}
+
+#[must_use = "page table updates will not be reflected until the changed pages are flushed from the TLB"]
+pub struct Handle<'a, L: level::PointsToPage> {
+    page: Page<VAddr, L::Size>,
+    entry: &'a mut Entry<L>,
+}
+
+/// An x86_64 page table.
+#[repr(align(4096))]
+#[repr(C)]
+pub struct PageTable<L> {
+    entries: [Entry<L>; ENTRIES],
+}
+
+#[derive(Clone)]
+#[repr(transparent)]
+pub struct Entry<L> {
+    entry: u64,
+    _level: PhantomData<L>,
+}
+
+#[tracing::instrument(level = "info")]
+pub fn init_paging(vm_offset: VAddr) {
+    VM_OFFSET.store(vm_offset.as_usize(), Ordering::Release);
+
+    tracing::info!("initializing paging...");
+
+    let (pml4_page, flags) = crate::control_regs::cr3::read();
+    tracing::debug!(?pml4_page, ?flags);
+    tracing::trace!("old PML4:");
+    let pml4 = PageTable::<level::Pml4>::current(vm_offset);
+
+    // Log out some details about our starting page table.
+    let mut present_entries = 0;
+    for (idx, entry) in (&pml4.entries[..]).iter().enumerate() {
+        if entry.is_present() {
+            tracing::trace!(idx, ?entry);
+            present_entries += 1;
+        }
+    }
+    tracing::trace!(present_entries);
+
+    // Create the recursive pagetable entry.
+    assert!(
+        !pml4.entries[RECURSIVE_INDEX].is_present(),
+        "bootloader must not have used entry 511"
+    );
+    pml4.entries[RECURSIVE_INDEX]
+        .set_present(true)
+        .set_writable(true)
+        .set_phys_addr(pml4_page.base_address());
+    tracing::info!("recursive entry created");
+
+    // Forcibly flush the entire TLB by resetting cr3.
+    unsafe {
+        crate::control_regs::cr3::write(pml4_page, flags);
+    }
+    tracing::info!("new PML4 set");
+
+    // Log out some details about our starting page table.
+    let mut present_entries = 0;
+    for (idx, entry) in (&pml4.entries[..]).iter().enumerate() {
+        if entry.is_present() {
+            tracing::trace!(idx, ?entry);
+            present_entries += 1;
+        }
+    }
+    tracing::trace!(present_entries);
+}
+
+/// This value should only be set once, early in the kernel boot process before
+/// we have access to multiple cores. So, technically, it could be a `static
+/// mut`. But, using an atomic is safe, even though it's not strictly necessary.
+static VM_OFFSET: AtomicUsize = AtomicUsize::new(core::usize::MAX);
+
+impl PageTable<level::Pml4> {
+    fn current(vm_offset: VAddr) -> &'static mut Self {
+        let (phys, _) = crate::control_regs::cr3::read();
+        unsafe { Self::from_pml4_page(vm_offset, phys) }
+    }
```

Comment:
It occurs to me that this should probably be `unsafe`, because there's no way to guarantee exclusivity, and you could pass in an arbitrary `vm_offset` value.

Reply:
i was unsure about this one given the "creating a possiblly-invalid pointer is safe, dereferencing it is unsafe" heuristic but yeah.

Reply:
Yeah, locking is probably a good call.